### ISSUE TITLE
[BUU] Change name, SKU and price of my products and variants

### ISSUE THREAD

**Issue by mariocarabotta** [openfoodfoundation/openfoodnetwork#11059]
## Context
This issue introduces the ability to edit some of the fields for products and variants. We start with a subset of text fields.

## Out of scope
- [BUU] Change producer, unit type, category and tax category #11060
- [BUU] Change if a product inherits properties or not #11063
- [BUU] Change stock levels for my products #11062
- [BUU] Change unit number and unit display #11061

## Description

**- As an:** enterprise user 
**- On page:** `/admin/products`
**- I want to be able to:** edit name and SKU for products, name, SKU and price for variants
**- So that:** I can provide the right information to customers for sale

## Acceptance Criteria & Tests

**Scenario 1: Hovering**
Given that I am in `/admin/products`
And my catalogue has loaded
When I hover on a product or variant
Then I see that name, SKU, and price are editable

**Scenario 2: Active fields**
Given that I am hovering on a product or variant
When I select an editable field
Then that field becomes active
And I can change its content

**Scenario 3: Editing first product**
Given that I have selected an editable field
When I change its content
Then I see a message stating that 1 product has changed

**Scenario 4: Editing multiple rows for first product**
Given that I have already edited one field for one product
When I edit a second row for that same product (for example editing 2 different variants for the same product)
Then the message still states that only one product has changed

**Scenario 5: Editing multiple products**
Given that I have edited one product only
When I edit more products
Then the message tells me the number of products I have edited

**Scenario 6: Discard changes**
Given I have edited products in my catalogue
And I'm seeing the message stating how many products have changed
When I discard the changes
Then the table reverts to its original state
And the message disappears

**Scenario 7: Request for saving**
Given I have edited products in my catalogue
And I'm seeing the message stating how many products have changed
When I save the changes
Then my request is sent
And I see a loading indicator
And I can't interact with the table interface

**Scenario 8: Save successful**
Given I have requested to save updates to my products
And I'm seeing a loading indicator
When the products are saved successfully
Then the table refreshes with the new data
And I temporarily see a success message

**Scenario 9: Save failed - single product**
Given I have requested to save updates to one of my products
And I have not filled a product or a variant name
And I'm seeing a loading indicator
When the system has tried to save and failed
Then I see an general error message telling me that my product could not be saved
And I see a specific error message highlighting the field that has caused the error

**Scenario 10: Save failed - multiple products**
Given I have requested to save updates to more than one of my products
And I have not filled a product or a variant name for multiple products
And I'm seeing a loading indicator
When the system has tried to save and failed
Then I see an general error message telling me the number of products that could not be saved could not be saved
And I see a specific error message highlighting the fields that have caused the error

**Scenario 11: Save failed - Some saved and some failed**
Given I have requested to save updates to more than one of my products
And I have not filled a product or a variant name for some of those products
And I'm seeing a loading indicator
When the system has tried to save and partially failed
Then I see an general error message telling me that some products were saved, and some were not
And I see a specific error message highlighting the fields that have caused the error

## Design specs
A prototype for the success flow is available here (change name of first product, change price of carrot variant, then save) > https://www.figma.com/proto/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery?page-id=363%3A2039&type=design&node-id=392-2584&viewport=376%2C497%2C0.08&t=dSLOjkxd3SHsINta-1&scaling=min-zoom&mode=design

A prototype for the partial failure flow is available here (try to save, edit name for second product with error, save again) > 
https://www.figma.com/proto/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery---Back-Office?page-id=363%3A2039&type=design&node-id=588-3256&viewport=297%2C95%2C0.03&t=wakt53xrS03QXHDY-1&scaling=min-zoom&show-proto-sidebar=1&starting-point-node-id=588%3A3256&mode=design

Figma screens are available here > https://www.figma.com/file/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery?type=design&node-id=363%3A2039&mode=design&t=XHz1hiwJaN6WPkoz-1

**New components and styles**

| Component | Example |
| --- | --- |
| Hover state for table rows | ![Screenshot 2023-07-14 at 8 22 30](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/0d926455-1b15-43f5-9ee2-acf36acba660) |
| Regular text field |![Screenshot 2023-07-14 at 8 23 23](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/d210e935-e728-4ab9-8e04-de379759fdd4)|
|Persistent message| ![Screenshot 2023-07-14 at 8 25 10](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/ecd84f64-7d2c-44e1-ae1f-aa6c94383a8f)|
|Regular transparent button|![Screenshot 2023-07-14 at 8 26 06](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/da4c65ce-3932-4d19-a515-401586e6237a)|
|Regular primary button|![Screenshot 2023-07-14 at 8 27 26](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/74e78c44-d1bc-4218-b621-3d41eea83b21)|
|Overlaying loading indicator|![Screenshot 2023-07-14 at 8 28 09](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/42a881df-23d9-4aae-a54e-08e033abaf45)|
|Overlaying success message|![Screenshot 2023-07-14 at 8 28 15](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/f7253987-6f80-4c48-bf67-664fc5b74411)|
|Field inline error|![Screenshot 2023-08-16 at 14 46 19](https://github.com/openfoodfoundation/openfoodnetwork/assets/6696729/89eb9cb8-5200-4914-8d16-3f6ef22cf9ab)|

**Comment by dacook** [comment]
* Ellipsis (…): This will hide when a field is hovered/active/error state.  (@dacook to investigate how to implement).

Also noting that we are intending to allow batching of updates (edit mutiple then click save once). Just like current screen.

**Comment by mariocarabotta** [comment]
note from conversation with @dacook - display price without currency symbol, always display 2 decimals (@dacook to explore if using dot or comma)

**Comment by dacook** [comment]
@mariocarabotta, I just realised the old bulk products screen doesn't do this:

Given I have edited products in my catalogue
And I'm seeing the message stating how many products have changed
* **When I navigate to another page** (eg click browser back button, a link in the navigation menu, the 'new product' button); or
* **When I choose another search filter,**

Then a warning message appears, asking to confirm losing unsaved changes before leaving.

**Comment by dacook** [comment]
In short, for the first example we'd have to have an unstyled popup like this. You can't customise the message:
![Screen Shot 2023-08-01 at 4 45 07 pm](https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/41bb238b-d8e0-4e4a-9aec-b6303e1ed070)

But for choosing another search filter we could style it, eg:
![Screen Shot 2023-08-01 at 4 40 12 pm](https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/a29ba306-cce6-494f-b295-cc2513ffb839)
(I think we could do this with any other links on the page too, but I'm unsure at this stage if it's a good idea.)

**Comment by mariocarabotta** [comment]
after discussing with @dacook, the default browser dialog will be used for every action that triggers going to a new page.
"When I choose another search filter" scenario is already handled by disabling the search and filtering with an overlay (see prototype)

**Comment by mariocarabotta** [comment]
list of possible errors:

Product Name can’t be blank
Variant Unit Name can’t be blank
Count on hand must be greater than or equal to 0

slack thread here > https://openfoodnetwork.slack.com/archives/C01CXQNJ1J6/p1690355249724539

@dacook is implementing saving so that:
- when multiple products are edited, the saving process saves each individual product independently
- if it encounters one of the above issues, it will keep going with the saving process but not save the ones with issues
- once the saving process is done, we can identify which products have been saved and which have not, and display errors on each field that didn't pass validation
- even if products change name and the catalogue list is ordered by name, once a product has been saved, it will stay in the current catalogue page until the whole interface gets refreshed, so that people can see its updates

**Comment by mariocarabotta** [comment]
prototype for errors management is [here for review](https://www.figma.com/proto/ddL6h7H9It5ZmUuVYQxzAD/Product-List?page-id=481%3A16918&type=design&node-id=610-5722&viewport=334%2C214%2C0.03&t=qhYgsgFUuScsZ1EP-1&scaling=min-zoom&starting-point-node-id=610%3A5722&mode=design)

if we implement these enhancements as part of the current issue, we can close https://github.com/openfoodfoundation/openfoodnetwork/issues/11190

**Comment by mariocarabotta** [comment]
updated with more detailed error scenarios and designs for error states

**Comment by mariocarabotta** [comment]
scenarios 7 and 8 are covered in this separate card

[Saving and success message tweaks](https://github.com/openfoodfoundation/openfoodnetwork/issues/11517)

**Comment by dacook** [comment]
Hey @mariocarabotta, I can't remember if we talked about this before, but did we have a plan for the "x" icon here?

I just reviewed the state of our FontAwesome, and we're on v3.2.1 which doesn't have this icon. It looks like we discussed upgrading to v6 but couldn't without paying for it. I just started looking at intermediate versions, and v5 does have an icon like this (https://fontawesome.com/v5/icons/times-circle?f=classic&s=solid&sz=xs&pc=%23c85136), so maybe it's worth looking into that..

**Comment by dacook** [comment]
Thanks for confirming the icon Mario, I can now see it here: https://fontawesome.com/v3/icon/remove-sign

**Comment by dacook** [comment]
Following on from https://github.com/openfoodfoundation/openfoodnetwork/issues/11059#issuecomment-1667035475, we discussed how to handle variants. We will consider variants as part of the product, so:

1. Each variant is validated with the parent product. If anything is invalid, nothing is saved. IE: 
   * if there is an error on the product, the variants remain unsaved
   * if there is an error in any of the variants, the other variants and product remain unsaved
 2. Product is saved once all variants are validated
 3. Variants are saved once product is valid and saved

In any case, I'll try to make the unsaved values will be clearly marked with orange (or red if error) border so that it's clear what is unsaved.
